File: crates/saber-tools/src/write.rs
//! `write`: creates or overwrites files inside the writable roots. Existing
//! targets must have been read first (session-state guard); writes are
//! atomic (tmp + rename) so crashes never leave half files.

use crate::ToolContext;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
pub struct WriteParams {
    pub path: String,
    pub content: String,
}

pub async fn run_write(ctx: &ToolContext, params: WriteParams) -> (String, bool) {
    match run_write_inner(ctx, params).await {
        Ok(summary) => (summary, false),
        Err(e) => (format!("write failed: {e}"), true),
    }
}

async fn run_write_inner(ctx: &ToolContext, params: WriteParams) -> Result<String, String> {
    let path = PathBuf::from(&params.path);
    let resolved = ctx.policy.check_write(&path).map_err(|e| e.to_string())?;
    let existed = resolved.exists();
    if existed && !ctx.has_been_read(&resolved) {
        return Err(format!(
            "{} already exists; read it before overwriting with write",
            params.path
        ));
    }
    if let Some(parent) = resolved.parent() {
        // Parent must stay inside the writable roots (mkdir -p is fine,
        // but the resolved target was already policy-checked).
        tokio::fs::create_dir_all(parent)
            .await
            .map_err(|e| format!("mkdir {}: {e}", parent.display()))?;
    }
    let _guard = ctx.file_lock(&resolved).await;
    let tmp = resolved.with_extension(format!("sabertmp-{}", std::process::id()));
    tokio::fs::write(&tmp, &params.content)
        .await
        .map_err(|e| format!("{}: {e}", params.path))?;
    tokio::fs::rename(&tmp, &resolved)
        .await
        .map_err(|e| format!("{}: {e}", params.path))?;
    Ok(format!(
        "{} {} ({} bytes)",
        if existed { "overwrote" } else { "created" },
        params.path,
        params.content.len()
    ))
}

File: packages/tools/src/write.ts
import * as fs from "node:fs/promises";
import * as path from "node:path";
import { checkWrite } from "./path-policy.ts";

export interface WriteParams { path: string; content: string }

export async function runWrite(
  policy: Parameters<typeof checkWrite>[0],
  readFiles: Set<string>,
  params: WriteParams,
): Promise<{ content: string; isError: boolean }> {
  const denied = checkWrite(policy, params.path);
  if (denied) return { content: `write failed: ${denied}`, isError: true };
  const resolved = path.resolve(params.path);
  const existed = await fs.stat(resolved).then(() => true, () => false);
  if (existed && !readFiles.has(resolved)) {
    return { content: `write failed: ${params.path} exists; read it first`, isError: true };
  }
  await fs.mkdir(path.dirname(resolved), { recursive: true });
  // Atomic write: tmp + rename
  const tmp = resolved + `.sabertmp-${process.pid}-${Date.now()}`;
  await fs.writeFile(tmp, params.content);
  if (existed) {
    const stat = await fs.stat(resolved);
    await fs.chmod(tmp, stat.mode);
  }
  await fs.rename(tmp, resolved);
  return {
    content: `${existed ? "overwrote" : "created"} ${params.path} (${params.content.length} bytes)`,
    isError: false,
  };
}
